DEVPROD-13970 Add s3 rest route for s3.get and s3.put commands

ZackarySantana opened the pull request:
DEVPROD-13970

### Description
This adds an s3 rest route that generates credentials for s3.get and s3.put targeting internal buckets.

The credentials are denied permission to everything. The way the route generates credentials as a subset of evergreen's is by calling AssumeRole using the same role as the Evergreen's IRSA role.

### Testing
Unit testing (when the project command is implemented will do e2e testing).

ablack12 commented on cloud/sts.go:
```diff
@@ -19,6 +19,8 @@ import (
 	// AssumeRole gets the credentials for a role as the given task.
 	AssumeRole(ctx context.Context, taskID string, opts AssumeRoleOptions) (AssumeRoleCredentials, error)
+	// GetCallerIdentity gets the credentials for a role as the given task.
+	GetCallerIdentity(ctx context.Context) (string, error)
```
I think it's a mistake that this description is the same as AssumeRole, especially since it doens't take in a taskID. Can you re-clarify? How does this function differ from AssumeRole?

ZackarySantana replied:
Mentioned above but renamed to say the Caller Identity's ARN

ablack12 commented on rest/route/agent.go:
```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
+// This route is used to retrieve credentials for an S3 bucket for a task.
+// s3.put and s3.get call this route when the command is targetting an
+// internal bucket.
+type awsS3 struct {
```
Do we actually need a different route for this also? This seems the same as the assume_role route, but with an additional step if callerARN isn't provided, and the bucket. Should we consider just adding some boolean logic to the existing route?

ZackarySantana replied:
This route acts a bit different. `callerARN` isn't user-provided and is from Evergreen's IRSA role. This route is creating credentials that have restricted permissions to s3 but it AssumeRole Evergreen's IRSA role.

ablack12 replied:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.

ablack12 replied:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.

ablack12 replied:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.

ZackarySantana commented on apimodels/agent_models.go:
```diff
@@ -233,6 +233,29 @@ type AssumeRoleResponse struct {
+	return catcher.Resolve()
+}
+
+// S3Response the credentials for accessing an S3 bucket.
```
Updated!

ZackarySantana commented on cloud/sts.go:
```diff
@@ -19,6 +19,8 @@ import (
 	// AssumeRole gets the credentials for a role as the given task.
 	AssumeRole(ctx context.Context, taskID string, opts AssumeRoleOptions) (AssumeRoleCredentials, error)
+	// GetCallerIdentity gets the credentials for a role as the given task.
+	GetCallerIdentity(ctx context.Context) (string, error)
```
Updated!

ZackarySantana commented on cloud/sts.go:
```diff
@@ -93,6 +95,21 @@ func (s *stsManagerImpl) AssumeRole(ctx context.Context, taskID string, opts Ass
 	}, nil
 }
 
+// GetCallerIdentityARN gets the ARN of the caller identity.
```
Switched to the new name!

ZackarySantana commented on rest/route/agent.go:
```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
 	})
 }
+
+// POST /rest/v2/task/{task_id}/aws/s3
```
We can do s3_credentials

ZackarySantana commented on rest/route/agent.go:
```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err, "finding task '%s'", h.taskID))
+	}
+	if t == nil {
```
Yes. It's going to be used in that ticket to generate the session policy. I rephrase the comment above the fake session policy to mention `create session policy based off of provided task`

ZackarySantana commented on rest/route/agent.go:
```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
+// This route is used to retrieve credentials for an S3 bucket for a task.
+// s3.put and s3.get call this route when the command is targetting an
+// internal bucket.
+type awsS3 struct {
```
Closing as dupe (smh github)

ZackarySantana commented on rest/route/agent.go:
```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
 }
+
+// POST /rest/v2/task/{task_id}/aws/s3
+// This route is used to retrieve credentials for an S3 bucket for a task.
```
Sure!

ZackarySantana commented:
I renamed the AssumeRole and s3 credentials response to `AWSCredentials`. I don't expect the output to change for either route (it it comes, we can split the responses). I also renamed some of the structs.

ablack12 commented on cloud/ec2_client.go:
```diff
@@ -1052,6 +1055,26 @@ func (c *awsClientImpl) AssumeRole(ctx context.Context, input *sts.AssumeRoleInp
+		ctx,
+		func() (bool, error) {
+			msg := makeAWSLogMessage("GetCallerIdentity", fmt.Sprintf("%T", c), input)
+			// GetCallerIdentity doesn't require permissions and we should retry on
```
this says we _should_ retry on any error, but here we return false, which would cause us to not retry. Should we be returning true? Alternatively if we don't intend to retry, then we don't need the retry loop at all, we can just call GetCallerIdentity directly

ZackarySantana replied:
Added an if err != nil -> return true to retry on error and adjusted the comment to reflect the intention (retry on network errors)

ablack12 reviewed (approved):
LGTM with last comment
